### ISSUE TITLE
Rustdoc: Very long doc compile times (95%+ taken by render_html)

### ISSUE THREAD

**Issue by JustusFluegel** [rust-lang/rust#132775]
I've recently tried to build rustdocs in one of our projects, and I've noticed some very odd 1h+ compile times for the docs (see the attached profile) with a majority of the time taken by the render_html item in the flamegraph. Also, during that time only a single core is pinned to 100%, even though multiple cores are available. (I don't know if rustdoc is multithreaded in that stage so if it is not that should be as expected)

My question now is: Is this expected behavior? If not, or if you need more information, I can try to create a minimum example that is able to reproduce this issue if that is something someone wants :)

![Image](https://github.com/user-attachments/assets/bffcd308-43c6-40e9-a629-6155547b74a8)

[prof file](https://drive.google.com/file/d/1gE3GFRhPKDnTq4n3-6h0NeJBxtC9xVEE/view?usp=sharing)

(since the .mm_profdata file is too big I uploaded it to google drive instead :) )

**Comment by clubby789** [comment]
Code to reproduce the issue is needed to see what's going on. Is it a large project with lots of documentation? Long render times seems somewhat expected if that's the case

**Comment by JustusFluegel** [comment]
Fair enough, I'll try to provide that that next week, probably on Monday (need to strip out a bit of code I am not allowed to share here )  :)  Thanks for the reply anyways , Have a nice weekend!

**Comment by aDotInTheVoid** [comment]
Everything being in 1 thread is entirely unremarkable. The code taking 1h to document is concerning (assuming rustc runs in a reasonable time).

https://github.com/aDotInTheVoid/everybody_loops might be useful to minimize it, as _generally_ rustdoc doesn’t care about function bodies.

**Comment by GuillaumeGomez** [comment]
It does care about function bodies because you might have `impl` blocks inside them:

```rust
pub struct Bar;

fn foo() {
    impl Bar {
        // This is callable everywhere and should be documented.
        pub fn yeay() {}
    }
}
```

It's why we removed this optimization. The same is true for all blocks unfortunately.
